src/components/validation/validators/component_spec.rs

```diff
@@ -68,15 +68,135 @@ impl Validator for ComponentSpecValidator {
             }
         }
 
-        // TODO: Check for the relevant telemetry events for the given component type.
-
-        Ok(vec![
+        let mut run_out = vec![
             format!(
                 "sent {} inputs and received {} outputs",
                 inputs.len(),
                 outputs.len()
             ),
             format!("received {} telemetry events", telemetry_events.len()),
-        ])
+        ];
+
+        let mut out = validate_telemetry(component_type, inputs, outputs, telemetry_events)?;
+        run_out.append(&mut out);
+
+        Ok(run_out)
+    }
+}
+
+fn validate_telemetry(
+    component_type: ComponentType,
+    inputs: &[TestEvent],
+    _outputs: &[Event],
+    telemetry_events: &[Event],
+) -> Result<Vec<String>, Vec<String>> {
+    let mut out: Vec<String> = Vec::new();
+    let mut errs: Vec<String> = Vec::new();
+
+    match component_type {
+        ComponentType::Source => {
+            match &mut validate_component_events_received(inputs, telemetry_events) {
+                Err(e) => errs.append(e),
+                Ok(m) => out.append(m),
+            }
+        }
+        _ => {}
+    }
+
+    if errs.is_empty() {
+        Ok(out)
+    } else {
+        Err(errs)
+    }
+}
+
+fn validate_component_events_received(
+    inputs: &[TestEvent],
+    telemetry_events: &[Event],
+) -> Result<Vec<String>, Vec<String>> {
+    let mut errs: Vec<String> = Vec::new();
+
+    let mut metrics = Vec::<Metric>::new();
+    for t in telemetry_events.iter() {
+        if let vector_core::event::Event::Metric(m) = t {
+            if m.name() == SourceMetrics::ComponentEventsReceived.name() {
+                if let Some(tags) = m.tags() {
+                    if tags.get("component_name").unwrap_or("") == "test_source" {
```

Comment:
We probably want to lift `"test_source"` to a constant, and replace instances of it in the validation topology builder with the constant as well, just so we don't end up with unexpected drift down the line.